### PR TITLE
RENAME: some enhancements

### PR DESCRIPTION
**Pull request by tom-wa**
added:
renaming of keys
toupper conversion
tolower conversion

### CODE REVIEW

File: src/plugins/rename/rename.c [hunk 1]
```diff
@@ -17,33 +17,121 @@
...
+	if(replaceWith != NULL)
+		addToLen = strlen(replaceWith);
 
-	/* marks the position in the key name where the retained parts start */
-	size_t afterCut = afterParentKey + cutPathLen;
+	size_t maxNewLength = strlen(keyName(key)) + addToLen;
 
-	if (keyNameLen < afterCut) return 0;
+	char *newName = malloc(maxNewLength+1);
+	memset(newName, 0, maxNewLength+1);
```
Comment (markus2330) [inline src/plugins/rename/rename.c]: elektraCalloc exactly does this.

File: src/plugins/rename/rename.c [hunk 2]
```diff
@@ -17,33 +17,121 @@
...
+	free(returnName);
+}
 
-	/* marks the position in the key name where the parent key path ends */
-	size_t afterParentKey = strlen(keyName(parentKey)) + 1;
+Key *elektraKeyCreateNewName(const Key *key, const Key *parentKey, const char *cutPath, const char *replaceWith, const char *toUpperPath, const char *toLowerPath)
+{
+	size_t addToLen = 0;
+	if(replaceWith != NULL)
+		addToLen = strlen(replaceWith);
```
Comment (markus2330) [inline src/plugins/rename/rename.c]: prefer to use {} here

File: src/plugins/rename/rename.c [hunk 3]
```diff
@@ -17,33 +17,121 @@
...
 {
-	size_t cutPathLen = strlen(cutPath);
-	size_t keyNameLen = strlen(keyName(key));
+	int (*conversion)(int);
+	if(toCase == TOUPPER)
+		conversion = toupper;
+	else
+		conversion = tolower;
+	char *returnName = malloc(strlen(newName)+1);
+	memset(returnName, 0, strlen(newName)+1);
```
Comment (markus2330) [inline src/plugins/rename/rename.c]: elektraCalloc exactly does this

File: src/plugins/rename/rename.c [hunk 4]
```diff
@@ -17,33 +17,121 @@
...
 char *keyNameGetOneLevel(const char *name, size_t *size);
 
-Key *elektraKeyCutNamePart(const Key *key, const Key *parentKey, const char *cutPath)
+static void doConversion(char *newName, int levels, int toCase)
 {
-	size_t cutPathLen = strlen(cutPath);
-	size_t keyNameLen = strlen(keyName(key));
+	int (*conversion)(int);
+	if(toCase == TOUPPER)
+		conversion = toupper;
```
Comment (markus2330) [inline src/plugins/rename/rename.c]: please use {} or ? for a one-liner

**Comment by markus2330** [comment]
@tom-wa nicely done!

@strahlex @fberlakovich can you take a look at this? (ideally try it out)

File: src/plugins/rename/rename.c [hunk 5]
```diff
@@ -17,33 +17,121 @@
...
 
 #define ELEKTRA_ORIGINAL_NAME_META "origname"
-
+#define TOLOWER (-1)
+#define TOUPPER 1
 // TODO defined privately in keyhelpers.c, API break possible..
 char *keyNameGetOneLevel(const char *name, size_t *size);
 
-Key *elektraKeyCutNamePart(const Key *key, const Key *parentKey, const char *cutPath)
+static void doConversion(char *newName, int levels, int toCase)
```
Comment (markus2330) [inline src/plugins/rename/rename.c]: There are two new coding style rules: space after method names and spaces around `*`.
So the line would be:
`static void doConversion (char * newName, int levels, int toCase)`

File: src/plugins/rename/testmod_rename.c [hunk 6]
```diff
@@ -4,7 +4,7 @@
  * \brief A plugin that makes use of libaugeas to read and write configuration files
  *
  * \copyright BSD License (see doc/COPYING or http://www.libelektra.org)
- *
+ org9*
```
Comment (markus2330) [inline src/plugins/rename/testmod_rename.c]: ?

File: src/plugins/rename/rename.c [hunk 7]
```diff
@@ -114,9 +222,9 @@ int elektraRenameGet(Plugin *handle, KeySet *returned, Key *parentKey)
 	if (!strcmp (keyName(parentKey), "system/elektra/modules/rename"))
 	{
 		KeySet *info =
-			#include "contract.h"
+#include "contract.h"
 
-		ksAppend (returned, info);
+			ksAppend (returned, info);
```
Comment (markus2330) [inline src/plugins/rename/rename.c]: ?

**Comment by strahlex** [comment]
How exactly can this plugin be used?
I tried:
```bash
$ kdb mount `pwd`/mkwrapper.ini system/machinekit/cnc ini dbus type rename
$ kdb set /tolower 0
$ kdb get /machinekit/cnc/DISPLAY/DISPLAY
mkwrapper
$ kdb get /machinekit/cnc/display/display
Did not find key
$ kdb get /tolower
0
```

File: src/plugins/rename/README.md [hunk 8]
```diff
@@ -54,9 +54,28 @@ after the parent key path.
...
+
+Using the `/toupper` or `/tolower` global configuration key, or the `rename/toupper` or `rename/tolower` MetaTag the rename plugin will 
+convert the keynames to uppercase or lowercase.
+The supplied values tell the plugin how many levels starting from the right will be converted. `toupper` and `tolower` can be combined.
+When no value or "0" is supplied with the keys the whole name below the parentkey will be converted.
+
+### EXAMPLE ###
+```
+/toupper = 1
+/tolower = 3
```
Comment (markus2330) [inline src/plugins/rename/README.md]: please give full `kdb mount` example!

Comment (markus2330) [inline src/plugins/rename/README.md]: would be nice to get this example, then we can finish this PR

or do you want us to directly test the new INI?

**Comment by strahlex** [comment]
Well done, works for me now.

**Pull request merged**